core/src/main/java/com/scalar/db/transaction/consensuscommit/RollbackMutationComposer.java

```diff
@@ -77,26 +80,39 @@ private Put composePut(Operation base, TransactionResult result) throws Executio
     TransactionTableMetadata metadata = tableMetadataManager.getTransactionTableMetadata(base);
     LinkedHashSet<String> beforeImageColumnNames = metadata.getBeforeImageColumnNames();
 
-    Map<String, Value<?>> map = new HashMap<>();
+    List<Column<?>> columns = new ArrayList<>();
     result
-        .getValues()
+        .getColumns()
         .forEach(
             (k, v) -> {
               if (beforeImageColumnNames.contains(k)) {
                 String key = k.substring(Attribute.BEFORE_PREFIX.length());
-                map.put(key, v.copyWith(key));
+                if (key.equals(Attribute.VERSION) && v.getIntValue() == 0) {
+                  columns.add(IntColumn.ofNull(Attribute.VERSION));
+                } else {
+                  columns.add(v.copyWith(key));
+                }
               }
             });
 
-    return new Put(result.getPartitionKey().get(), result.getClusteringKey().orElse(null))
-        .forNamespace(base.forNamespace().get())
-        .forTable(base.forTable().get())
-        .withCondition(
-            new PutIf(
-                new ConditionalExpression(ID, toIdValue(id), Operator.EQ),
-                new ConditionalExpression(STATE, toStateValue(result.getState()), Operator.EQ)))
-        .withConsistency(Consistency.LINEARIZABLE)
-        .withValues(map.values());
+    PutBuilder.Buildable putBuilder =
+        Put.newBuilder()
+            .namespace(base.forNamespace().get())
+            .table(base.forTable().get())
+            .partitionKey(result.getPartitionKey().get())
+            .condition(
+                ConditionBuilder.putIf(ConditionBuilder.column(ID).isEqualToText(id))
```

Comment:
I'm wondering if `version` should be contained in this conditions. It doesn't need to be?

Reply:
It's enough to check only the transaction ID and the status to ensure that no other transaction updates this record for either commit or rollback. Moreover, I think only transaction ID would be mandatory here since it is decided that this record must be rolled back (no chance of role forwarding), and the transaction ID always be updated if someone recovers.

Reply:
Ah, sorry for the silly question. I had a wrong idea that `ID` meant a key not a transaction ID at that moment. I sometimes confuse them...